### PR TITLE
Translate js/ts strings outside of a React context

### PR DESCRIPTION
**Pull request by azeier**
## List of all js/ts files
Relevant files were identified via `find -follow | grep '[^(\.d)]\.[tj]s$'`

File|Translations
---|---
./static/scripts/src/global.d.ts|-
./static/scripts/src/extractors.ts|-
./static/scripts/src/JoustEmbedder.ts|Joust loading error - **NOT IMPLEMENTED**
./static/scripts/src/TourManager.ts|Buttons
./static/scripts/src/GameFilters.ts|-
./static/scripts/src/CardData.ts|-
./static/scripts/src/utils/account.ts|-
./static/scripts/src/utils/collection.ts|-
./static/scripts/src/utils/enums.ts|-
./static/scripts/src/contants.ts|-
./static/scripts/src/entries/export-react.ts|-
./static/scripts/src/entries/polyfills.ts|-
./static/scripts/src/UserData.ts|-
./static/scripts/src/components/tables/cardtable/RowDataGenerator.ts|-
./static/scripts/src/components/tables/cardtable/CardTableColumns.ts|Fixed stub t() call
./static/scripts/src/i18n.ts|-
./static/scripts/src/RavenWatcher.ts|-
./static/scripts/src/metrics/InfluxMetricsBackend.ts|-
./static/scripts/src/metrics/MetricsBackend.d.ts|-
./static/scripts/src/metrics/GoogleAnalytics.ts|-
./static/scripts/src/metrics/BatchingMiddleware.ts|-
./static/scripts/src/metrics/MetricsReporter.ts|-
./static/scripts/src/DataManager.ts|-
./static/scripts/src/helpers.ts|Set names, CardClass names, "Other" archetype, ChartScheme names
./static/scripts/src/Settings.ts|-
./templates/joust_include.js|-
./templates/footerjs.js|-

## locale/en/frontend.json diff
```diff
diff --git a/en/frontend.json b/en/frontend.json
index 74823c8..8c08517 100644
--- a/en/frontend.json
+++ b/en/frontend.json

diff --git a/en/frontend.json b/en/frontend.json
index 74823c8..41968f5 100644
--- a/en/frontend.json
+++ b/en/frontend.json
@@ -23,6 +23,7 @@
   "Blizzard China (CN)…": "Blizzard China (CN)…",
   "Debug": "Debug",
   "Sign in with password…": "Sign in with password…",
+  "Other {cardClass}": "Other {cardClass}",
   "Not enough games for meaningful matchup data available.": "Not enough games for meaningful matchup data available.",
   "Core cards": "Core cards",
   "Popular cards": "Popular cards",
@@ -375,6 +376,9 @@
   "Previous season": "Previous season",
   "Latest expansion": "Latest expansion",
   "Latest patch": "Latest patch",
+  "All": "All",
+  "Taverns of Time": "Taverns of Time",
+  "Dream": "Dream",
   "My financial situation has changed": "My financial situation has changed",
   "I prefer to manually renew my subscriptions": "I prefer to manually renew my subscriptions",
   "Not enough support for Wild": "Not enough support for Wild",
@@ -755,5 +759,6 @@
   "Your deck tracker is too old!": "Your deck tracker is too old!",
   "We no longer support uploads from the version of Hearthstone you are using. If you are already on the latest version, please contact us for assistance.": "We no longer support uploads from the version of Hearthstone you are using. If you are already on the latest version, please contact us for assistance.",
   "Your replay is still processing. Check back soon!": "Your replay is still processing. Check back soon!",
-  "Something went wrong generating this replay. We're on it.": "Something went wrong generating this replay. We're on it."
+  "Something went wrong generating this replay. We're on it.": "Something went wrong generating this replay. We're on it.",
+  "Done": "Done"

```
## Remaining work
- Is there anything obvious I missed?
- I'm unsure how to proceed with `JoustEmbedder`. `embed()` is called from `replay_embed`, which does not have any React context either. Suggestions?

### CODE REVIEW

File: hsreplaynet/static/scripts/src/components/text/PrettyCardClass.tsx [hunk 1]
```diff
@@ -22,7 +22,7 @@ class PrettyCardClass extends React.Component<Props> {
 			!i18n.hasResourceBundle(i18n.language, "hearthstone") ||
 			!UserData.hasFeature("frontend-translations")
 		) {
-			return getHeroClassName(getCardClassName(cardClass));
+			return toTitleCase(getCardClassName(cardClass));
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/components/text/PrettyCardClass.tsx]: Definitely not. This should use the actual class name (using the `hearthstone` namespace). `toTitleCase` should essentially never be used.

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 2]
```diff
@@ -599,7 +610,33 @@ export function getHeroCardId(cardClass: CardClass | string): string | null {
...
+			return t("GLOBAL_CLASS_PRIEST");
+		case "ROGUE":
+			return t("GLOBAL_CLASS_ROGUE");
+		case "SHAMAN":
+			return t("GLOBAL_CLASS_SHAMAN");
+		case "WARLOCK":
+			return t("GLOBAL_CLASS_WARLOCK");
+		case "WARRIOR":
+			return t("GLOBAL_CLASS_WARRIOR");
+		case "DREAM":
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: I think it's safe to leave that one out?

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 3]
```diff
@@ -599,7 +610,33 @@ export function getHeroCardId(cardClass: CardClass | string): string | null {
 	return null;
 }
 
-export function getHeroClassName(cardClass: string): string {
+export function getHeroClassName(cardClass: string, t: TranslationFunction): string {
+	switch(cardClass) {
+		case "DEATHKNIGHT":
+			return t("GLOBAL_CLASS_DEATHKNIGHT");
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: You can probably simply do:

```ts
return t(`GLOBAL_CLASS_${cardClass}`);
```

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 4]
```diff
@@ -142,134 +145,142 @@ const costScheme: ChartScheme = {
...
+export function getSetNames(t: TranslationFunction) {
+	return {
+		core: t("GLOBAL_CARD_SET_CORE"),
+		expert1: t("GLOBAL_CARD_SET_EXPERT1"),
+		hof: t("GLOBAL_CARD_SET_HOF"),
+		naxx: t("GLOBAL_CARD_SET_NAXX"),
+		gvg: t("GLOBAL_CARD_SET_GVG"),
+		brm: t("GLOBAL_CARD_SET_BRM"),
+		tgt: t("GLOBAL_CARD_SET_TGT"),
+		tb: t("Tavern Brawl"),
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: You can use `GLOBAL_TAVERN_BRAWL`

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 5]
```diff
@@ -142,134 +145,142 @@ const costScheme: ChartScheme = {
...
-	gangs: "Mean Streets of Gadgetzan",
-	ungoro: "Journey to Un'Goro",
-	icecrown: "Knights of the Frozen Throne",
-	lootapalooza: "Kobolds and Catacombs",
-	gilneas: "The Witchwood",
-	taverns_of_time: "Taverns of Time",
-};
+export function getSetNames(t: TranslationFunction) {
+	return {
+		core: t("GLOBAL_CARD_SET_CORE"),
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: Maybe out of scope of this PR but why is this all lowercase / not using the enum?

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 6]
```diff
@@ -142,134 +145,142 @@ const costScheme: ChartScheme = {
...
-	og: "Whispers of the Old Gods",
-	kara: "One Night in Karazhan",
-	gangs: "Mean Streets of Gadgetzan",
-	ungoro: "Journey to Un'Goro",
-	icecrown: "Knights of the Frozen Throne",
-	lootapalooza: "Kobolds and Catacombs",
-	gilneas: "The Witchwood",
-	taverns_of_time: "Taverns of Time",
-};
+export function getSetNames(t: TranslationFunction) {
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: Much more efficient: `getSetNames(setName, t): string`

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 7]
```diff
@@ -142,134 +145,142 @@ const costScheme: ChartScheme = {
...
-		fill: "rgba(199, 156, 110, 0.7)",
-		name: "Warrior",
-	},
-	other: {
-		stroke: "rgba(122, 122, 122, 1)",
-		fill: "rgba(122, 122, 122, 0.7)",
-		name: "Other",
-	},
-};
+function getClassColorScheme (t: TranslationFunction): ChartScheme {
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: Much more efficient: `getClassColorScheme(className, t): string`

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 8]
```diff
@@ -142,134 +145,142 @@ const costScheme: ChartScheme = {
...
-		stroke: "rgba(0, 112, 222, 0.9)",
-		name: "Spell",
-	},
-	weapon: {
-		fill: "rgba(196, 30, 59, 0.5)",
-		stroke: "rgba(196, 30, 59, 0.9)",
-		name: "Weapon",
-	},
-};
+function getCardtypeScheme (t: TranslationFunction): ChartScheme {
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: Same comments as below apply: Don't return an array, switch to enums/uppercase strings.

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 9]
```diff
@@ -142,134 +145,142 @@ const costScheme: ChartScheme = {
...
-		stroke: "rgba(163, 53, 238, 0.9)",
-		name: "Epic",
-	},
-	legendary: {
-		fill: "rgba(255, 128, 0, 0.5)",
-		stroke: "rgba(255, 128, 0, 0.9)",
-		name: "Legendary",
-	},
-};
+function getRarityScheme(t: TranslationFunction): ChartScheme {
```
Comment (jleclanche) [inline hsreplaynet/static/scripts/src/helpers.ts]: Same comments as below apply: Don't return an array, switch to enums/uppercase strings.

File: hsreplaynet/static/scripts/src/components/text/PrettyCardClass.tsx (hunk 1 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/components/text/PrettyCardClass.tsx]: This is the case where the localization is not loaded.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 2 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/helpers.ts]: /shrug might as well have it.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 3 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/helpers.ts]: As benedict mentioned in discord, not the way we should do it.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 5 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/helpers.ts]: I'm sure there was some (bad) reason. But yeah definitely not part of this PR.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 7 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/helpers.ts]: Doesn't work with the way the schemes are used.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 8 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/helpers.ts]: Doesn't work with the way the schemes are used.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 9 above)
Comment (azeier) [reply hsreplaynet/static/scripts/src/helpers.ts]: Doesn't work with the way the schemes are used.

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 10]
```diff
@@ -142,134 +145,159 @@ const costScheme: ChartScheme = {
...
+			return t("GLOBAL_CARD_SET_UNGORO");
+		case "icecrown":
+			return t("GLOBAL_CARD_SET_ICECROWN");
+		case "lootapalooza":
+			return t("GLOBAL_CARD_SET_LOOTAPALOOZA");
+		case "gilneas":
+			return t("GLOBAL_CARD_SET_GILNEAS");
+		case "taverns_of_time":
+			return t("Taverns of Time");
+	}
```
Comment (beheh) [inline hsreplaynet/static/scripts/src/helpers.ts]: Add a fallback string.

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 11]
```diff
@@ -142,134 +145,159 @@ const costScheme: ChartScheme = {
...
-	og: "Whispers of the Old Gods",
-	kara: "One Night in Karazhan",
-	gangs: "Mean Streets of Gadgetzan",
-	ungoro: "Journey to Un'Goro",
-	icecrown: "Knights of the Frozen Throne",
-	lootapalooza: "Kobolds and Catacombs",
-	gilneas: "The Witchwood",
-	taverns_of_time: "Taverns of Time",
-};
+export function getSetName(set: string, t: TranslationFunction) {
```
Comment (beheh) [inline hsreplaynet/static/scripts/src/helpers.ts]: Returns `: string`.

File: hsreplaynet/static/scripts/src/helpers.ts [hunk 12]
```diff
@@ -142,134 +145,159 @@ const costScheme: ChartScheme = {
...
+	return {
+		all: {
+			stroke: "rgba(169, 169, 169, 1)",
+			fill: "rgba(169, 169, 169, 0.7)",
+			name: t("All"),
+		},
+		neutral: {
+			stroke: "rgba(169, 169, 169, 1)",
+			fill: "rgba(169, 169, 169, 0.7)",
+			name: t("GLOBAL_CLASS_NEUTRAL"),
```
Comment (beheh) [inline hsreplaynet/static/scripts/src/helpers.ts]: Let's use `getHeroClassName` for these.

File: hsreplaynet/static/scripts/src/components/text/PrettyCardClass.tsx [hunk 13]
```diff
@@ -22,7 +22,7 @@ class PrettyCardClass extends React.Component<Props> {
...
 		}
 
 		switch (cardClass) {
```
Comment (beheh) [inline hsreplaynet/static/scripts/src/components/text/PrettyCardClass.tsx]: Use the new `helpers.getHeroClassName`.

File: hsreplaynet/static/scripts/src/helpers.ts (hunk 3 above)
Comment (beheh) [reply hsreplaynet/static/scripts/src/helpers.ts]: The usage is fine. You should explicitely state the namespace though (using `{ ns: "hearthstone" }`).

**Comment by beheh** [comment]
I'll still take a look at JoustEmbedder.

**Pull request merged**